File: cryptoheir-rs/src/contract.rs
//! CryptoHeir contract ABI encoding and bytecode loading

use crate::Result;
use alloy::{
    primitives::{Address, Bytes, U256},
    sol,
    sol_types::{SolCall, SolInterface},
};

/// Path to the compiled CryptoHeir contract artifact
const CONTRACT_ARTIFACT_PATH: &str = "../foundry/out/CryptoHeir.sol/CryptoHeir.json";

// Define the contract ABI by loading from Foundry's compiled artifacts
// This automatically syncs with the Solidity contract - just run `forge build` and recompile
// Note: sol! macro requires a string literal (can't use the constant above)
sol!(
    #[sol(rpc)]
    CryptoHeir,
    "../foundry/out/CryptoHeir.sol/CryptoHeir.json"
);

/// Load contract bytecode from Foundry artifacts
pub fn load_bytecode() -> Result<Bytes> {
    let artifact_str = std::fs::read_to_string(CONTRACT_ARTIFACT_PATH).map_err(|e| {
        eyre::eyre!(
            "Failed to read contract artifact at {}: {}. Make sure Foundry contracts are compiled.",
            CONTRACT_ARTIFACT_PATH,
            e
        )
    })?;

    let artifact: serde_json::Value = serde_json::from_str(&artifact_str)?;

    let bytecode_str = artifact["bytecode"]["object"]
        .as_str()
        .ok_or_else(|| eyre::eyre!("Bytecode not found in artifact"))?;

    let bytecode: Bytes = bytecode_str.parse()?;

    if bytecode.is_empty() {
        return Err(eyre::eyre!("Contract bytecode is empty"));
    }

    Ok(bytecode)
}

/// Encode deposit function call
/// Returns (calldata, value)
/// Note: Parameter order matches Solidity: token, beneficiary, amount, deadline
pub async fn encode_deposit(
    beneficiary: Address,
    amount: U256,
    deadline: u64,
    token: Option<Address>,
) -> Result<(Bytes, Option<U256>)> {
    let token_addr = token.unwrap_or(Address::ZERO);
    let deadline_u256 = U256::from(deadline);

    // If token is zero address, this is a native token deposit
    let value = if token.is_none() {
        Some(amount)
    } else {
        None
    };

    // Encode the function call - using actual parameter names from contract
    let call = CryptoHeir::depositCall {
        _token: token_addr,
        _beneficiary: beneficiary,
        _amount: amount,
        _deadline: deadline_u256,
    };

    let calldata = call.abi_encode().into();

    Ok((calldata, value))
}

/// Encode claim function call
pub fn encode_claim(id: U256) -> Result<Bytes> {
    let call = CryptoHeir::claimCall { _inheritanceId: id };
    Ok(call.abi_encode().into())
}

/// Encode reclaim function call
pub fn encode_reclaim(id: U256) -> Result<Bytes> {
    let call = CryptoHeir::reclaimCall { _inheritanceId: id };
    Ok(call.abi_encode().into())
}

/// Encode extendDeadline function call
pub fn encode_extend_deadline(id: U256, new_deadline: u64) -> Result<Bytes> {
    let call = CryptoHeir::extendDeadlineCall {
        _inheritanceId: id,
        _newDeadline: U256::from(new_deadline),
    };
    Ok(call.abi_encode().into())
}

/// Encode transferFeeCollector function call
pub fn encode_transfer_fee_collector(new_collector: Address) -> Result<Bytes> {
    let call = CryptoHeir::transferFeeCollectorCall { newFeeCollector: new_collector };
    Ok(call.abi_encode().into())
}

/// Encode acceptFeeCollector function call
pub fn encode_accept_fee_collector() -> Result<Bytes> {
    let call = CryptoHeir::acceptFeeCollectorCall {};
    Ok(call.abi_encode().into())
}

/// Decode contract revert data into a human-readable error message
pub fn decode_contract_error(revert_data: &Bytes) -> Option<String> {
    // Try to decode as a CryptoHeir error
    if let Ok(error) = CryptoHeir::CryptoHeirErrors::abi_decode(revert_data, false) {
        let message = match error {
            CryptoHeir::CryptoHeirErrors::InvalidBeneficiary(_) => {
                "Invalid beneficiary address (cannot be zero address or sender)"
            }
            CryptoHeir::CryptoHeirErrors::InvalidDeadline(_) => {
                "Invalid deadline (must be in the future)"
            }
            CryptoHeir::CryptoHeirErrors::InsufficientAmount(_) => {
                "Insufficient amount (cannot be zero)"
            }
            CryptoHeir::CryptoHeirErrors::InvalidTokenTransfer(_) => {
                "Invalid token transfer (msg.value must be zero for ERC20 deposits)"
            }
            CryptoHeir::CryptoHeirErrors::InheritanceNotFound(_) => {
                "Inheritance not found"
            }
            CryptoHeir::CryptoHeirErrors::AlreadyClaimed(_) => {
                "Inheritance already claimed"
            }
            CryptoHeir::CryptoHeirErrors::DeadlineNotReached(_) => {
                "Deadline not reached yet"
            }
            CryptoHeir::CryptoHeirErrors::DeadlineAlreadyPassed(_) => {
                "Deadline already passed"
            }
            CryptoHeir::CryptoHeirErrors::OnlyDepositor(_) => {
                "Only the depositor can perform this action"
            }
            CryptoHeir::CryptoHeirErrors::OnlyBeneficiary(_) => {
                "Only the beneficiary can perform this action"
            }
            CryptoHeir::CryptoHeirErrors::OnlyFeeCollector(_) => {
                "Only the fee collector can perform this action"
            }
            CryptoHeir::CryptoHeirErrors::InvalidFeeCollector(_) => {
                "Invalid fee collector address"
            }
            CryptoHeir::CryptoHeirErrors::NoPendingTransfer(_) => {
                "No pending fee collector transfer"
            }
            _ => "Unknown contract error"
        };
        Some(message.to_string())
    } else {
        None
    }
}

File: cryptoheir-go/internal/contract/contract.go
package contract

import (
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math/big"
	"os"
	"path/filepath"
	"strings"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
)

// ContractArtifact represents the Foundry contract artifact structure
type ContractArtifact struct {
	ABI      json.RawMessage `json:"abi"`
	Bytecode struct {
		Object string `json:"object"`
	} `json:"bytecode"`
}

var (
	contractABI      abi.ABI
	contractBytecode []byte
)

// Initialize loads the contract artifact and parses ABI
func Initialize() error {
	// Get the contract artifact path relative to project root
	artifactPath := filepath.Join("..", "foundry", "out", "CryptoHeir.sol", "CryptoHeir.json")

	// Read artifact file
	data, err := os.ReadFile(artifactPath)
	if err != nil {
		return fmt.Errorf("failed to read contract artifact: %w", err)
	}

	// Parse artifact
	var artifact ContractArtifact
	if err := json.Unmarshal(data, &artifact); err != nil {
		return fmt.Errorf("failed to parse contract artifact: %w", err)
	}

	// Parse ABI
	parsedABI, err := abi.JSON(strings.NewReader(string(artifact.ABI)))
	if err != nil {
		return fmt.Errorf("failed to parse contract ABI: %w", err)
	}
	contractABI = parsedABI

	// Parse bytecode (remove 0x prefix if present)
	bytecodeHex := strings.TrimPrefix(artifact.Bytecode.Object, "0x")
	bytecode, err := hex.DecodeString(bytecodeHex)
	if err != nil {
		return fmt.Errorf("failed to decode contract bytecode: %w", err)
	}
	contractBytecode = bytecode

	return nil
}

// LoadBytecode returns the contract deployment bytecode
func LoadBytecode() ([]byte, error) {
	if len(contractBytecode) == 0 {
		return nil, fmt.Errorf("contract not initialized, call Initialize() first")
	}
	return contractBytecode, nil
}

// EncodeDeposit encodes the deposit function call
// deposit(address _token, address _beneficiary, uint256 _amount, uint256 _deadline)
func EncodeDeposit(beneficiary common.Address, amount *big.Int, deadline *big.Int, token *common.Address) ([]byte, *big.Int, error) {
	if contractABI.Methods == nil {
		return nil, nil, fmt.Errorf("contract not initialized, call Initialize() first")
	}

	// Default token to zero address (native ETH)
	tokenAddr := common.Address{}
	if token != nil {
		tokenAddr = *token
	}

	// Pack the function call
	data, err := contractABI.Pack("deposit", tokenAddr, beneficiary, amount, deadline)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to encode deposit: %w", err)
	}

	// If native token (zero address), return the amount as value
	var value *big.Int
	if tokenAddr == (common.Address{}) {
		value = amount
	}

	return data, value, nil
}

// EncodeClaim encodes the claim function call
// claim(uint256 _inheritanceId)
func EncodeClaim(inheritanceID *big.Int) ([]byte, error) {
	if contractABI.Methods == nil {
		return nil, fmt.Errorf("contract not initialized, call Initialize() first")
	}

	data, err := contractABI.Pack("claim", inheritanceID)
	if err != nil {
		return nil, fmt.Errorf("failed to encode claim: %w", err)
	}

	return data, nil
}

// EncodeReclaim encodes the reclaim function call
// reclaim(uint256 _inheritanceId)
func EncodeReclaim(inheritanceID *big.Int) ([]byte, error) {
	if contractABI.Methods == nil {
		return nil, fmt.Errorf("contract not initialized, call Initialize() first")
	}

	data, err := contractABI.Pack("reclaim", inheritanceID)
	if err != nil {
		return nil, fmt.Errorf("failed to encode reclaim: %w", err)
	}

	return data, nil
}

// EncodeExtendDeadline encodes the extendDeadline function call
// extendDeadline(uint256 _inheritanceId, uint256 _newDeadline)
func EncodeExtendDeadline(inheritanceID *big.Int, newDeadline *big.Int) ([]byte, error) {
	if contractABI.Methods == nil {
		return nil, fmt.Errorf("contract not initialized, call Initialize() first")
	}

	data, err := contractABI.Pack("extendDeadline", inheritanceID, newDeadline)
	if err != nil {
		return nil, fmt.Errorf("failed to encode extendDeadline: %w", err)
	}

	return data, nil
}

// EncodeTransferFeeCollector encodes the transferFeeCollector function call
// transferFeeCollector(address newFeeCollector)
func EncodeTransferFeeCollector(newCollector common.Address) ([]byte, error) {
	if contractABI.Methods == nil {
		return nil, fmt.Errorf("contract not initialized, call Initialize() first")
	}

	data, err := contractABI.Pack("transferFeeCollector", newCollector)
	if err != nil {
		return nil, fmt.Errorf("failed to encode transferFeeCollector: %w", err)
	}

	return data, nil
}

// EncodeAcceptFeeCollector encodes the acceptFeeCollector function call
// acceptFeeCollector()
func EncodeAcceptFeeCollector() ([]byte, error) {
	if contractABI.Methods == nil {
		return nil, fmt.Errorf("contract not initialized, call Initialize() first")
	}

	data, err := contractABI.Pack("acceptFeeCollector")
	if err != nil {
		return nil, fmt.Errorf("failed to encode acceptFeeCollector: %w", err)
	}

	return data, nil
}

// DecodeContractError attempts to decode a contract revert error
func DecodeContractError(revertData []byte) string {
	if len(revertData) < 4 {
		return "unknown error"
	}

	// Try to match error selector
	for name, abiError := range contractABI.Errors {
		if len(revertData) >= 4 {
			errorID := revertData[:4]
			// Compare first 4 bytes (selector)
			if len(abiError.ID) >= 4 &&
				errorID[0] == abiError.ID[0] &&
				errorID[1] == abiError.ID[1] &&
				errorID[2] == abiError.ID[2] &&
				errorID[3] == abiError.ID[3] {
				// Found matching error selector
				return fmt.Sprintf("contract error: %s", name)
			}
		}
	}

	// Check for standard Error(string) revert
	errorSig := []byte{0x08, 0xc3, 0x79, 0xa0} // Error(string) selector
	if len(revertData) >= 4 &&
		revertData[0] == errorSig[0] &&
		revertData[1] == errorSig[1] &&
		revertData[2] == errorSig[2] &&
		revertData[3] == errorSig[3] {
		// Try to decode the string (skip selector, decode ABI-encoded string)
		if len(revertData) > 4 {
			return fmt.Sprintf("revert: %s", string(revertData[4:]))
		}
	}

	return fmt.Sprintf("unknown error: 0x%x", revertData[:min(len(revertData), 32)])
}

// min returns the smaller of two ints
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
